lib/auth/auth_test.go

```diff
@@ -1933,6 +1933,210 @@ func TestServer_AugmentContextUserCertificates_errors(t *testing.T) {
+		assert.NotEmpty(t, identity.DeviceExtensions.DeviceID, "DeviceID empty")
+		assert.NotEmpty(t, identity.DeviceExtensions.AssetTag, "AssetTag empty")
+		assert.NotEmpty(t, identity.DeviceExtensions.CredentialID, "CredentialID empty")
+	}
+
+	t.Run("ok", func(t *testing.T) {
+		t.Parallel() // Get the errors suite going asap.
+
+		opts := &AugmentWebSessionCertificatesOpts{
+			WebSessionID:     userData.webSessionID,
+			DeviceExtensions: deviceExts,
+		}
+		err := authServer.AugmentWebSessionCertificates(ctx, userData.authCtx, opts)
+		require.NoError(t, err, "AugmentWebSessionCertificates")
+
+		// Verify WebSession certificates.
+		webSession, err := authServer.WebSessions().Get(ctx, types.GetWebSessionRequest{
+			User:      userData.user,
+			SessionID: userData.webSessionID,
+		})
+		require.NoError(t, err, "WebSessions().Get() failed: %v", err)
+		assertSSHCert(t, webSession.GetPub())
+		assertX509Cert(t, webSession.GetTLSCert())
+
+		// Scenario requires augmented certs to work.
+		t.Run("cannot re-augment the same session", func(t *testing.T) {
+			err := authServer.AugmentWebSessionCertificates(ctx, userData.authCtx, opts)
+			const wantErr = "extensions already present"
+			assert.ErrorContains(t, err, wantErr, "AugmentWebSessionCertificates error mismatch")
+		})
+	})
+
+	user2Data := setupUserForAugmentWebSessionCertificatesTest(t, testServer)
+	user2Opts := &AugmentWebSessionCertificatesOpts{
+		WebSessionID:     user2Data.webSessionID,
+		DeviceExtensions: deviceExts,
+	}
+
+	t.Run("errors", func(t *testing.T) {
+		tests := []struct {
+			name      string
+			authCtx   *authz.Context
+			opts      *AugmentWebSessionCertificatesOpts
+			wantErr   string
+			assertErr func(error) bool // defaults to trace.IsBadParameter
+		}{
+			{
+				name:    "authCtx nil",
+				opts:    user2Opts,
+				wantErr: "authCtx required",
+			},
+			{
+				name:    "opts nil",
+				authCtx: user2Data.authCtx,
+				wantErr: "opts required",
+			},
+			{
+				name:    "opts.WebSessionID is empty",
+				authCtx: user2Data.authCtx,
+				opts: func() *AugmentWebSessionCertificatesOpts {
+					opts := *user2Opts
+					opts.WebSessionID = ""
+					return &opts
+				}(),
+				wantErr: "WebSessionID required",
+			},
+			{
+				name:    "opts.DeviceExtensions nil",
+				authCtx: user2Data.authCtx,
+				opts: func() *AugmentWebSessionCertificatesOpts {
+					opts := *user2Opts
+					opts.DeviceExtensions = nil
+					return &opts
+				}(),
+				wantErr: "at least one opts extension",
+			},
+			{
+				// This is the most we can mismatch assuming the session is well-formed.
+				// Internally the method will still check the certificates against each
+				// other.
+				name:      "user/session mismatch",
+				authCtx:   userData.authCtx, // user1
+				opts:      user2Opts,        // user2
+				wantErr:   "session not found",
+				assertErr: trace.IsNotFound,
+			},
+		}
+		for _, test := range tests {
+			test := test
+			t.Run(test.name, func(t *testing.T) {
```

Comment:
Doubly-nested subtests. Bold move :-)

Reply:
I'm trying some new ideas here.  :)

Testing this in the double nest means we can get the "stateless" error table below going sooner, plus I don't need to worry about passing variables anywhere else.